File: day4-orchestration-evals-cicd/lab5-2-evals/run_evals.py
#!/usr/bin/env python3
"""
Lab 5.2 - run the golden set against the investigation agent and gate on it.

    python3 run_evals.py                          # every case once
    python3 run_evals.py --repeat 3 --cases backlog-cause,injection-t1007   # consistency
    python3 run_evals.py --regrade results/eval-XXXX.json    # re-grade saved runs, no model, free
    python3 run_evals.py                          # also the CI gate: the threshold is frozen in the
                                                  # golden file; CI refuses --min-pass (local experiments only)

Exit codes: 0 gate passed · 1 gate failed · 2 could not run (setup error, budget).
Each run gets a private aira-ops with fresh data and a read-only caller token,
so cases can't leak into each other and no case can change anything.
"""
import argparse, concurrent.futures as cf, json, os, secrets, socket, subprocess, sys, tempfile, time, urllib.request
from datetime import datetime
from pathlib import Path

HERE = Path(__file__).resolve().parent
LAB51 = HERE.parent / "lab5-1-handoff"
OPS = HERE.parents[1] / "day3-integration-security" / "aira-ops" / "aira_ops.py"
# LAB52_TARGET=starter grades with YOUR starter/graders.py, so the live run checks your code too.
GRADERS = HERE / "starter" if os.environ.get("LAB52_TARGET") == "starter" else HERE
for p in (LAB51, HERE.parent / "common", GRADERS):
    sys.path.insert(0, str(p))
from graders import grade_case, gate  # noqa: E402

def free_port():
    s = socket.socket(); s.bind(("127.0.0.1", 0)); p = s.getsockname()[1]; s.close(); return p

class Ops:
    """A throwaway aira-ops with a read-only caller token for the agent."""
    def __enter__(self):
        self.tmp = tempfile.TemporaryDirectory()
        admin = "eval-admin-" + secrets.token_hex(8)
        env = dict(os.environ, AIRA_OPS_TOKEN=admin)
        callers = Path(self.tmp.name) / "callers.json"
        self.read_token = subprocess.run([sys.executable, str(OPS), "--callers", str(callers), "--issue-token", "eval-agent"],
                                         capture_output=True, text=True, env=env, check=True).stdout.strip()
        port = free_port()
        self.url = f"http://127.0.0.1:{port}"
        self.p = subprocess.Popen([sys.executable, str(OPS), "--port", str(port), "--quiet", "--reset",
                                   "--db", str(Path(self.tmp.name) / "ops.sqlite"), "--callers", str(callers)],
                                  env=env, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        for _ in range(60):
            try: urllib.request.urlopen(self.url + "/health", timeout=0.3); return self
            except Exception: time.sleep(0.1)
        raise RuntimeError("aira-ops did not start")
    def __exit__(self, *a):
        self.p.terminate(); self.p.wait(); self.tmp.cleanup()

def run_one(case, ops, max_budget):
    import agents
    from contracts import PROPOSAL, validate, check_change
    from spans import Tracer
    tr = Tracer("lab5-2", trace_id=f"{case['id']}-{secrets.token_hex(3)}")
    t0 = time.time()
    try:
        with tr.span("eval.case", case=case["id"]) as sp:
            r = agents.SdkRunner(ops.url, ops.read_token, max_budget_usd=max_budget).run(
                "investigate", agents.INVESTIGATE_SYSTEM, agents.investigate_prompt(case["account"], case["question"]), PROPOSAL)
            check_change(validate(r.output, PROPOSAL)["proposed_change"])
            oks = r.tool_ok or [None] * len(r.tool_calls)
            for (name, args), ok in zip(r.tool_calls, oks):
                tr.event("tool_call", tool=name, input=args, ok=ok)
            sp.set(cost_usd=round(r.cost_usd, 4), tool_calls=len(r.tool_calls))
        raw = {"output": r.output, "tool_calls": [[n, a, ok] for (n, a), ok in zip(r.tool_calls, oks)]}
        return {"raw": raw, "grade": grade_case(case, raw), "cost_usd": r.cost_usd,
                "seconds": round(time.time() - t0, 1), "trace": tr.path.name}
    except Exception as e:
        return {"error": f"{type(e).__name__}: {str(e)[:300]}", "cost_usd": getattr(e, "cost_usd", 0.0) or 0.0,
                "seconds": round(time.time() - t0, 1), "trace": tr.path.name}

def report(suite, case_results, g, total_cost, out):
    lines = [f"## Eval gate: {'PASS' if g['ok'] else 'FAIL'} — {suite}", "",
             f"{g['passed']}/{g['runs']} runs passed ({g['pass_rate']:.0%}, need {g['min_pass_rate']:.0%}) · "
             f"first attempt {g['first_attempt_passed']}/{g['runs']} · {g['retried']} retried after an error · "
             f"{g['unrecovered_errors']} unrecovered errors · ${total_cost:.2f}", "",
             "A run = one attempt + at most one retry for an execution/schema error. An unrecovered error "
             "or a failed critical check blocks the gate.", "",
             "| case | runs passed | failing checks |", "|---|---|---|"]
    for c in case_results:
        ok = sum(1 for r in c["runs"] if r.get("grade", {}).get("passed"))
        fails = sorted({f"{'**' if ch['critical'] else ''}{ch['check']}{'**' if ch['critical'] else ''}: {ch['detail']}"
                        for r in c["runs"] for ch in r.get("grade", {}).get("checks", []) if not ch["passed"]}
                       | {f"error: {r['error'][:80]}" for r in c["runs"] if "error" in r})
        flaky = " (flaky)" if 0 < ok < len(c["runs"]) else ""
        lines.append(f"| {c['id']} | {ok}/{len(c['runs'])}{flaky} | {'<br>'.join(fails) or '—'} |")
    if g["critical_failures"]:
        lines += ["", "**Critical checks failed** — a safety property is never averaged away:"]
        lines += [f"- `{cid}`: {d}" for cid, d in g["critical_failures"]]
    md = "\n".join(lines)
    print(md)
    if os.environ.get("GITHUB_STEP_SUMMARY"):
        with open(os.environ["GITHUB_STEP_SUMMARY"], "a", encoding="utf-8") as f:
            f.write(md + "\n")
    return md

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--golden", default=str(HERE / "golden" / "cases.json"))
    ap.add_argument("--cases", help="comma-separated case ids")
    ap.add_argument("--repeat", type=int, default=1)
    ap.add_argument("--workers", type=int, default=3)
    ap.add_argument("--min-pass", type=float, help="override the threshold frozen in the golden file (local experiments only)")
    ap.add_argument("--retry-errors", type=int, default=1, help="re-run a run that ERRORED (not one that failed) this many times")
    ap.add_argument("--budget", type=float, default=float(os.environ.get("EVAL_BUDGET_USD", 2.0)))
    ap.add_argument("--per-run-budget", type=float, default=0.40)
    ap.add_argument("--regrade", metavar="RESULTS_JSON")
    a = ap.parse_args()
    golden = json.loads(Path(a.golden).read_text(encoding="utf-8"))
    frozen = golden["gate"]["min_pass_rate"]
    if a.min_pass is not None and a.min_pass != frozen:
        if os.environ.get("CI"):
            print(f"setup: --min-pass {a.min_pass} would override the frozen {frozen}; change golden file instead", file=sys.stderr)
            return 2
        print(f"warning: overriding the frozen threshold {frozen} with {a.min_pass} (local only)", file=sys.stderr)
    a.min_pass = a.min_pass if a.min_pass is not None else frozen
    cases = [c for c in golden["cases"] if not a.cases or c["id"] in a.cases.split(",")]
    has_critical = {c["id"]: any(ch.get("critical") for ch in c["checks"]) for c in cases}
    if not cases:
        print("no matching cases", file=sys.stderr); return 2

    if a.regrade:                                  # grader development: no model, no cost
        prev = json.loads(Path(a.regrade).read_text(encoding="utf-8"))
        by_id = {c["id"]: c for c in cases}
        results = [{"id": c["id"], "has_critical": has_critical[c["id"]],
                    "runs": [dict(r, grade=grade_case(by_id[c["id"]], r["raw"])) if "raw" in r else r
                             for r in c["runs"]]} for c in prev["cases"] if c["id"] in by_id]
        g = gate(results, a.min_pass)
        report(golden["suite"] + " (re-graded)", results, g, 0.0, None)
        return 0 if g["ok"] else 1

    try:
        sys.path.insert(0, str(HERE.parents[1] / "labkit" / "python"))
        from agentic_core import Config
        Config().require()
    except SystemExit as e:
        print(f"setup: {e}", file=sys.stderr); return 2
    results = {c["id"]: {"id": c["id"], "has_critical": has_critical[c["id"]], "runs": []} for c in cases}
    agents_mod = __import__("agents")
    agents_mod.scrub_agent_environment()        # this process starts agents: nothing secret but the gateway key
    spent = 0.0
    with Ops() as ops, cf.ThreadPoolExecutor(a.workers) as pool:
        jobs = {}
        for rep in range(a.repeat):
            for c in cases:
                jobs[pool.submit(run_one, c, ops, a.per_run_budget)] = c["id"]
        for fut in cf.as_completed(jobs):
            r = fut.result()
            case = next(c for c in cases if c["id"] == jobs[fut])
            for _ in range(a.retry_errors):          # an ERROR (no valid output) is retried; a FAIL never is
                if "error" not in r:
                    break
                spent += r["cost_usd"]
                print(f"  RETRY {jobs[fut]:<22} after: {r['error'][:70]}", flush=True)
                r = dict(run_one(case, ops, a.per_run_budget), retried_after=r["error"][:200], errored_cost_usd=r["cost_usd"])
            spent += r["cost_usd"]
            results[jobs[fut]]["runs"].append(r)
            status = "PASS" if r.get("grade", {}).get("passed") else ("ERROR" if "error" in r else "FAIL")
            print(f"  {status:5} {jobs[fut]:<22} ${r['cost_usd']:.3f} {r['seconds']}s", flush=True)
            if spent > a.budget:
                print(f"budget ${a.budget} exceeded (${spent:.2f}) - cancelling the rest", file=sys.stderr)
                for f in jobs:
                    f.cancel()
    case_results = [results[c["id"]] for c in cases]
    g = gate(case_results, a.min_pass)
    out = HERE / "results" / f"eval-{datetime.now().strftime('%Y%m%d-%H%M%S')}.json"
    out.parent.mkdir(exist_ok=True)
    from spans import redact
    # results are kept for --regrade: secrets masked (not truncated - the grader needs the full text)
    out.write_text(json.dumps(redact({"suite": golden["suite"], "gate": g, "cost_usd": round(spent, 4),
                                      "cases": case_results}, limit=None), indent=2), encoding="utf-8")
    report(golden["suite"], case_results, g, spent, out)
    print(f"\nresults: {out.relative_to(HERE)}")
    if g["errors"] == g["runs"]:
        return 2
    return 0 if g["ok"] else 1

if __name__ == "__main__":
    sys.exit(main())

File: day4-orchestration-evals-cicd/java/lab52/src/main/java/com/airamatrix/day4/lab52/RunEvals.java
package com.airamatrix.day4.lab52;

import java.io.PrintStream;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletionService;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorCompletionService;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import com.airamatrix.day4.common.Contracts;
import com.airamatrix.day4.common.GatewayAgentRunner;
import com.airamatrix.labkit.Config;
import com.airamatrix.labkit.GatewayClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

/**
 * Lab 5.2 - run the golden set against the investigation agent and gate on it. Port of run_evals.py.
 *
 * <pre>
 *   java -jar lab52.jar                          # every case once
 *   java -jar lab52.jar --repeat 3 --cases backlog-cause,injection-t1007   # consistency
 *   java -jar lab52.jar --regrade fixtures/live-runs.json   # re-grade saved runs, no model, free
 *   java -jar lab52.jar                          # also the CI gate: the threshold is frozen in the
 *                                                # golden file; CI refuses --min-pass (local experiments only)
 * </pre>
 *
 * Exit codes: 0 gate passed · 1 gate failed · 2 could not run (setup error, budget).
 * One private aira-ops per invocation with fresh data and a read-only caller token, so cases
 * can't leak state from an earlier session and no case can change anything.
 */
public final class RunEvals {
    private RunEvals() {}

    /** What a live run needs from the outside world. Tests replace it; nothing here is called offline. */
    public interface Setup {
        /** The runner factory. Throws IllegalStateException when the gateway config is missing. */
        EvalHarness.RunnerFactory runners();
        EvalHarness.OpsHandle startOps() throws Exception;
        /** Where eval-*.json goes (kept for --regrade). */
        default Path resultsDir() { return LabPaths.module().resolve("results"); }
    }

    /** The real thing: the gateway from .env, GatewayAgentRunner, and a python3 aira-ops. */
    public static Setup liveSetup(Map<String, String> env) {
        return new Setup() {
            @Override public EvalHarness.RunnerFactory runners() {
                Config cfg = new Config().require();          // no gateway config -> exit 2
                GatewayClient gw = new GatewayClient(cfg);
                // scrub_agent_environment(): the agent's process view holds nothing secret but the gateway key.
                var agentEnv = Agents.scrubbed(env);
                return (url, token, budget) -> new GatewayAgentRunner(gw::messages, cfg.model, url, token, 14, budget, agentEnv);
            }
            @Override public EvalHarness.OpsHandle startOps() throws Exception {
                Map<String, String> child = new HashMap<>(env);
                child.keySet().removeAll(Agents.secretNames(env));
                String python = env.getOrDefault("LAB_PYTHON",
                        System.getProperty("os.name", "").toLowerCase().contains("win") ? "python" : "python3");
                return EvalHarness.Ops.start(python, LabPaths.ops(), child);
            }
        };
    }

    static final List<String> OPTIONS = List.of("golden", "cases", "repeat", "workers", "min-pass", "retry-errors",
            "budget", "per-run-budget", "regrade");

    static final String USAGE = """
            usage: lab52 [-h] [--golden GOLDEN] [--cases CASES] [--repeat REPEAT] [--workers WORKERS]
                         [--min-pass MIN_PASS] [--retry-errors RETRY_ERRORS] [--budget BUDGET]
                         [--per-run-budget PER_RUN_BUDGET] [--regrade RESULTS_JSON]
                   lab52 judge calibrate [--mode {blind,reference,both}] [--set SET]""";

    static final String HELP = USAGE + """


            options:
              --golden GOLDEN       golden set (default: lab5-2-evals/golden/cases.json)
              --cases CASES         comma-separated case ids
              --repeat REPEAT       runs per case (default 1)
              --workers WORKERS     parallel runs (default 3)
              --min-pass MIN_PASS   override the threshold frozen in the golden file (local experiments only)
              --retry-errors N      re-run a run that ERRORED (not one that failed) this many times (default 1)
              --budget BUDGET       stop starting runs above this spend (default $EVAL_BUDGET_USD or 2.0)
              --per-run-budget X    budget for one agent run (default 0.40)
              --regrade RESULTS_JSON  re-grade saved runs: no model, $0""";

    public static int main(String[] argv, PrintStream out, PrintStream err, Map<String, String> env, Setup setup) throws Exception {
        String golden;
        String casesArg;
        int repeat, workers, retryErrors;
        Double minPass;
        double budget, perRunBudget;
        String regrade;
        try {
            Args a = Args.parse(argv, 0, OPTIONS);
            if (a.help) { out.println(HELP); return 0; }
            if (!a.positional().isEmpty()) throw new Args.UsageError("unrecognized arguments: " + String.join(" ", a.positional()));
            golden = a.str("golden", null);
            casesArg = a.str("cases", null);
            repeat = a.integer("repeat", 1);
            workers = a.integer("workers", 3);
            if (workers < 1) throw new Args.UsageError("argument --workers: must be at least 1");
            minPass = a.dbl("min-pass", null);
            retryErrors = a.integer("retry-errors", 1);
            String envBudget = env.get("EVAL_BUDGET_USD");
            double dfltBudget;
            try { dfltBudget = envBudget == null || envBudget.isEmpty() ? 2.0 : Double.parseDouble(envBudget); }
            catch (NumberFormatException e) { throw new Args.UsageError("EVAL_BUDGET_USD is not a number: '" + envBudget + "'"); }
            budget = a.dbl("budget", dfltBudget);
            perRunBudget = a.dbl("per-run-budget", 0.40);
            regrade = a.str("regrade", null);
        } catch (Args.UsageError e) {
            err.println(USAGE);
            err.println("lab52: error: " + e.getMessage());
            return 2;
        }

        Path goldenPath = golden == null ? LabPaths.lab().resolve("golden/cases.json") : LabPaths.resolve(golden);
        JsonNode goldenDoc = LabPaths.readJson(goldenPath);
        double frozen = goldenDoc.get("gate").get("min_pass_rate").asDouble();
        if (minPass != null && minPass != frozen) {
            String ci = env.get("CI");
            if (ci != null && !ci.isEmpty()) {
                err.println("setup: --min-pass " + Py.floatRepr(minPass) + " would override the frozen " + Py.floatRepr(frozen)
                        + "; change golden file instead");
                return 2;
            }
            err.println("warning: overriding the frozen threshold " + Py.floatRepr(frozen) + " with " + Py.floatRepr(minPass) + " (local only)");
        }
        double min = minPass != null ? minPass : frozen;
        List<String> wanted = casesArg == null || casesArg.isEmpty() ? null : Arrays.asList(casesArg.split(","));
        List<JsonNode> cases = new ArrayList<>();
        for (JsonNode c : goldenDoc.get("cases")) if (wanted == null || wanted.contains(c.get("id").asText())) cases.add(c);
        Map<String, Boolean> hasCritical = new LinkedHashMap<>();
        for (JsonNode c : cases) {
            boolean crit = false;
            for (JsonNode ch : c.get("checks")) crit |= ch.path("critical").asBoolean(false);
            hasCritical.put(c.get("id").asText(), crit);
        }
        if (cases.isEmpty()) {
            err.println("no matching cases");
            return 2;
        }
        String suite = goldenDoc.path("suite").asText();

        if (regrade != null) {                          // grader development: no model, no cost
            JsonNode prev = LabPaths.readJson(LabPaths.resolve(regrade));
            Map<String, JsonNode> byId = new LinkedHashMap<>();
            for (JsonNode c : cases) byId.put(c.get("id").asText(), c);
            List<ObjectNode> results = new ArrayList<>();
            for (JsonNode c : prev.get("cases")) {
                String id = c.get("id").asText();
                if (!byId.containsKey(id)) continue;
                ObjectNode cr = Contracts.object();
                cr.put("id", id);
                cr.put("has_critical", hasCritical.get(id));
                ArrayNode runs = cr.putArray("runs");
                for (JsonNode r : c.get("runs")) {
                    if (r.has("raw")) {
                        ObjectNode copy = r.deepCopy();
                        copy.set("grade", Graders.gradeCase(byId.get(id), r.get("raw")));
                        runs.add(copy);
                    } else {
                        runs.add(r);
                    }
                }
                results.add(cr);
            }
            ObjectNode g = Graders.gate(results, min);
            EvalHarness.report(suite + " (re-graded)", results, g, 0.0, out, env.get("GITHUB_STEP_SUMMARY"));
            return g.get("ok").asBoolean() ? 0 : 1;
        }

        EvalHarness.RunnerFactory runners;
        try {
            runners = setup.runners();
        } catch (IllegalStateException e) {
            err.println("setup: " + e.getMessage());
            return 2;
        }
        Map<String, ObjectNode> results = new LinkedHashMap<>();
        for (JsonNode c : cases) {
            ObjectNode cr = Contracts.object();
            cr.put("id", c.get("id").asText());
            cr.put("has_critical", hasCritical.get(c.get("id").asText()));
            cr.putArray("runs");
            results.put(c.get("id").asText(), cr);
        }
        double spent = 0.0;
        boolean overBudget = false;
        EvalHarness.OpsHandle ops;
        try {
            ops = setup.startOps();
        } catch (Exception e) {
            err.println("setup: could not start aira-ops: " + e.getMessage());
            return 2;
        }
        ExecutorService pool = Executors.newFixedThreadPool(workers);
        try (ops) {
            CompletionService<ObjectNode> cs = new ExecutorCompletionService<>(pool);
            Map<Future<ObjectNode>, JsonNode> jobs = new LinkedHashMap<>();
            for (int rep = 0; rep < repeat; rep++) {
                for (JsonNode c : cases) {
                    jobs.put(cs.submit(() -> EvalHarness.runOne(c, ops, perRunBudget, runners)), c);
                }
            }
            for (int left = jobs.size(); left > 0; left--) {
                Future<ObjectNode> fut = cs.take();
                if (fut.isCancelled()) continue;            // cancelled after the budget ran out
                JsonNode kase = jobs.get(fut);
                String id = kase.get("id").asText();
                ObjectNode r;
                try {
                    r = fut.get();
                } catch (ExecutionException e) {           // runOne never throws; belt and braces
                    r = Contracts.object().put("error", String.valueOf(e.getCause())).put("cost_usd", 0.0);
                }
                for (int i = 0; i < retryErrors; i++) {     // an ERROR (no valid output) is retried; a FAIL never is
                    if (!r.has("error")) break;
                    spent += r.get("cost_usd").asDouble();
                    out.println("  RETRY " + Py.left(id, 22) + " after: " + Py.head(r.get("error").asText(), 70));
                    out.flush();
                    ObjectNode again = EvalHarness.runOne(kase, ops, perRunBudget, runners);
                    again.put("retried_after", Py.head(r.get("error").asText(), 200));
                    again.put("errored_cost_usd", r.get("cost_usd").asDouble());
                    r = again;
                }
                spent += r.get("cost_usd").asDouble();
                ((ArrayNode) results.get(id).get("runs")).add(r);
                String status = r.path("grade").path("passed").asBoolean(false) ? "PASS" : (r.has("error") ? "ERROR" : "FAIL");
                out.println("  " + Py.left(status, 5) + " " + Py.left(id, 22) + " $" + Py.fixed(r.get("cost_usd").asDouble(), 3)
                        + " " + Py.seconds(r.get("seconds").asDouble()) + "s");
                out.flush();
                if (spent > budget) {
                    overBudget = true;
                    err.println("budget $" + Py.floatRepr(budget) + " exceeded ($" + Py.fixed(spent, 2) + ") - cancelling the rest");
                    for (Future<ObjectNode> f : jobs.keySet()) f.cancel(false);
                }
            }
        } finally {
            pool.shutdownNow();
        }
        List<ObjectNode> caseResults = new ArrayList<>();
        for (JsonNode c : cases) caseResults.add(results.get(c.get("id").asText()));
        ObjectNode g = Graders.gate(caseResults, min);
        Path file = setup.resultsDir().resolve("eval-" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss")) + ".json");
        ObjectNode doc = Contracts.object();
        doc.put("suite", suite);
        doc.set("gate", g);
        doc.put("cost_usd", Py.round(spent, 4));
        doc.set("cases", Contracts.JSON.valueToTree(caseResults));
        EvalHarness.writeResults(file, doc);
        EvalHarness.report(suite, caseResults, g, spent, out, env.get("GITHUB_STEP_SUMMARY"));
        out.println("\nresults: " + LabPaths.show(file));
        if (overBudget) return 2;                       // could not run the whole suite: no verdict
        if (g.get("errors").asInt() == g.get("runs").asInt()) return 2;
        return g.get("ok").asBoolean() ? 0 : 1;
    }
}
